lib/fizzy/execution_context.hpp

```diff
@@ -4,38 +4,87 @@
 
 #pragma once
 
+#include "value.hpp"
+#include <cassert>
+#include <cstddef>
+
 namespace fizzy
 {
 /// The storage for information shared by calls in the same execution "thread".
 /// Users may decide how to allocate the execution context, but some good defaults are available.
 class ExecutionContext
 {
-    /// Call local execution context.
+    static constexpr size_t DefaultStackSpaceSegmentSize = 100;
+
+    /// Call depth increment guard.
     /// It will automatically decrement the call depth to the original value
     /// when going out of scope.
     class [[nodiscard]] LocalContext
     {
         ExecutionContext& m_shared_ctx;  ///< Reference to the shared execution context.
 
     public:
+        Value* stack_space = nullptr;
+        Value* prev_stack_space_segment = nullptr;
+        size_t prev_free_stack_space = 0;
+
         LocalContext(const LocalContext&) = delete;
         LocalContext(LocalContext&&) = delete;
         LocalContext& operator=(const LocalContext&) = delete;
         LocalContext& operator=(LocalContext&&) = delete;
 
-        explicit LocalContext(ExecutionContext& ctx) noexcept : m_shared_ctx{ctx}
+        LocalContext(ExecutionContext& ctx, size_t required_stack_space) : m_shared_ctx{ctx}
         {
             ++m_shared_ctx.depth;
+
+            prev_free_stack_space = m_shared_ctx.free_stack_space;
+
+            if (required_stack_space <= m_shared_ctx.free_stack_space)
+            {
+                // Must be a segment of default size or required_stack_space is 0.
+                const auto offset =
+                    DefaultStackSpaceSegmentSize - m_shared_ctx.free_stack_space;
+                stack_space = m_shared_ctx.stack_space_segment + offset;
+                prev_free_stack_space = m_shared_ctx.free_stack_space;
+                m_shared_ctx.free_stack_space -= required_stack_space;
+            }
+            else
+            {
+                prev_stack_space_segment = m_shared_ctx.stack_space_segment;
+                const auto new_segment_size =
+                    std::max(DefaultStackSpaceSegmentSize, required_stack_space);
+                m_shared_ctx.stack_space_segment = new Value[new_segment_size];
```

Comment:
It would allow a performance degrading behaviour like growing stack to almost `DefaultStackSpaceSegmentSize`, then calling a function in a loop that goes over the limit, allocates new segment and immediately deallocates after return, and that's repeated in a loop.
Should this be a concern? I can imagine mitigation with increasing allocation size each time, and keeping the last allocation size in `m_shared_ctx`.

Reply:
It is a concern, but the same as before (where stack space was dynamically allocated in some cases).
I don't think the increasing allocation size is the solution - it will still continue to deallocating new segment, now bigger each time.
The solution I considered is never deallocate segments. They work as linked list of buffers. However, such solution might require more housekeeping.